08-rails/forms-strong-params.md

```diff
@@ -0,0 +1,81 @@
+  else # save failed :(
+    render :new # show the new book form view again
+  end
+end
+```
+
+This code is written with the assumption that our `params` hash contains data as follows:
+```ruby
+{ book:
+  {
+    author: "some author",
+    title: "some title"
+  }
+}
+```
+
+Notice that the parameters passed into `Book.new` are all listed out individually. What if we wanted to pass all of the data from the `params` directly without listing out each individual field? What happens when we try: `@book = Book.new(params[:book)])`?
+
+What happened? Most likely, you've gotten a **ForbiddenAttributeError**.
+
+In Rails, controllers are **not permitted to pass parameters directly** from the form into our ActiveRecord model instances. This is a security feature to ensure users cannot update sensitive model attributes that they should not be updating.
+
+### Strong Parameters
+Rails has introduced the concept of **strong parameters** to use in our controllers to prevent these errors. Strong params will require us to give explicit access to a set of attributes that come through `params` that we'll then be able to pass to our model methods. We set up strong params once for our controller, and then we will no longer need to specify each field individually (in the way we had to above).
+
+### Update the Controller
+Now that we have an idea of _why_ we need strong params, let's see how to implement these in our controller.
+
+1. Create a `private` method section towards the bottom of our controller.
+2. Create a new method that we'll use to return the _permitted_ parameters.
+    ```ruby
+    private
+
+    def book_params
+      # code here
+    end
+    ```
+3. We'll use two special controller methods here which will allow us to define exactly which parameters (coming from the form) we want to allow. Those methods are `require` and `permit`.
+
+  We use `require` to identify the highest-level param which contains all others and that we cannot live without. We then use `permit` to list the params within that overall one which we want to allow.
+    ```ruby
+    private
+
+    def book_params
+      params.require(:book).permit(:author, :title)
+    end
+    ```
+4. We then use this method within our creation code above, in place of the code where we were writing out each parameter.
+    ```ruby
+    def create
+      @book = Book.new(book_params) #instantiate a new book using strong params
+      ...
+      # this code is the same
+    end
+    ```
+5. Run through the form submission again to ensure that we are no longer getting the `ForbiddenAttributeError`.
+
+A few things to note here:
+1. The new `book_params` method is using the `params` variable which the controller makes available to us (even though we're not passing it in as a parameter)
+2. We chain the method calls of `require` and `permit`. This works because the `:author` and `:title` fields are located **within** the `:book` param.
+
```

Comment:
Again, should be able to use `1.` for both of these and let the numbers happen automatically.